Generate a documentation page for core::mem::transmute.

jrvanwhy opened the pull request:
In `#[no_std]` environments, `std::mem::transmute` is unavailable. Searching for "core transmute" online only pulls up `core::intrinsics::transmute`, which is behind the (unstable) `core_intrinsics` feature flag. Users wishing to use transmute in `#[no_std]` environments typically should use `core::mem::transmute` instead, as it is stable. This documentation makes `core::mem::transmute` discoverable.

rust-highfive commented:
Thanks for the pull request, and welcome! The Rust team is excited to review your changes, and you should hear from @shepmaster (or someone else) soon.

If any changes to this PR are deemed necessary, please add them as extra commits. This ensures that the reviewer can see what has changed since they last reviewed the code. Due to the way GitHub handles out-of-date commits, this should also make it reasonably obvious what issues have or haven't been addressed. Large or tricky changes may require several passes of review and changes.

Please see [the contribution instructions](https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md) for more information.

lzutao commented:
I'm not a contributor. My thought is that when you search `transmute` in Rust doc, you would see all posible output.

![search](http://i.imgur.com/YDbQEao.png)

shepmaster commented:
> when you search `transmute` in Rust doc,

And (unfortunately in this case), the `intrinsics::` one appears before the one we'd want people to use (`mem::`).

I think this is fine, but I'd wonder how it would look for `std::intrinsics::transmute` — does it suggest using `core::mem`?

shepmaster commented:
Now, does this new fancy thing work yet...

r? @rust-lang/docs

shepmaster commented:
No, it does not.

r? @QuietMisdreavus

jrvanwhy commented:
> > when you search `transmute` in Rust doc,
> 
> And (unfortunately in this case), the `intrinsics::` one appears before the one we'd want people to use (`mem::`).

Actually, the one we want (`core::mem::transmute`) does not appear in the list at all. `core::mem::transmute_copy` does appear, as does `std::mem::transmute`. I suspect there's some mechanism trying to shepherd users towards `std` rather than `core` for items that exist in both, but that can be troublesome for `#![no_std]` users.

> I think this is fine, but I'd wonder how it would look for `std::intrinsics::transmute` — does it suggest using `core::mem`?

Unfortunately, the recommendation does show up in `std::intrinsics::transmute`'s docs. It also appears in `std::mem::transmute`'s docs. I'm not sure if it's possible to disable that -- if we can't prevent it, then this needs to be reworded.

shepmaster commented:
Hm. It's because `core::mem::transmute` is just a re-export:

https://github.com/rust-lang/rust/blob/19f8958f827e379be9b0ffc93fca580ec8ffac77/src/libcore/mem.rs#L17-L18

GuillaumeGomez commented:
You can (will be able to) filter your search (currently not in stable, only in nightly and beta...) so you won't have the search issue anymore. However, I approve the precision add!

scottmcm commented on src/libcore/intrinsics.rs:
```diff
@@ -729,6 +729,10 @@ extern "rust-intrinsic" {
     ///
+    /// `transmute` is re-exported by [core::mem](../mem/index.html) as
+    /// `core::mem::transmute`, which may be used without the `core_intrinsics`
+    /// feature flag.
```
These docs are going to show up for all four of `{core|std}::{mem|intrinsics}::transmute`, right?  It might be worth pointing at both locations, in that case.

QuietMisdreavus commented:
A better solution than adding the extra blurb in the docs is to apply `#[doc(inline)]` to [the `pub use intrinsics::transmute` statement](https://github.com/rust-lang/rust/blob/01f8e25b15f4ab157c8e7c9c56054df7595ec0e1/src/libcore/mem.rs#L17-L18) instead. This will make a new page for `core::mem::transmute` appear, which will also add it to the search index.

scottmcm commented:
Ooh, nice, @QuietMisdreavus!  That sounds great, and would let us make the docs for the intrinsic just be "don't use this; use it from core instead".

QuietMisdreavus commented:
> and would let us make the docs for the intrinsic just be "don't use this; use it from core instead"

Actually, the doc for the intrinsic is going to be the same in all four places it's available, because it's just re-exported around. There's no way to fix that, since rustdoc doesn't let you add different docs to a re-export.

jrvanwhy commented:
> A better solution than adding the extra blurb in the docs is to apply `#[doc(inline)]` to [the `pub use intrinsics::transmute` statement](https://github.com/rust-lang/rust/blob/01f8e25b15f4ab157c8e7c9c56054df7595ec0e1/src/libcore/mem.rs#L17-L18) instead. This will make a new page for `core::mem::transmute` appear, which will also add it to the search index.

Thank you! I'll give that a try -- it sounds like a much better solution.

jrvanwhy commented:
I added the doc page for core::mem::transmute. For my own use cases, this change would've been plenty, so I removed the additional paragraph from the doc comment. I can replace the comment if others feel that transmute is sufficiently unique to deserve a "recommended exports of" comment.

QuietMisdreavus commented:
Sorry for letting this sit for so long! Let's get this merged!

@bors r+ rollup

bors commented:
:pushpin: Commit 1c8c94af51e10e0593163d6a8ff41513271d7eab has been approved by `QuietMisdreavus`

<!-- @bors r=QuietMisdreavus 1c8c94af51e10e0593163d6a8ff41513271d7eab -->
